plugins/samples/hmac_authcookie/plugin.cc

```diff
@@ -15,45 +15,125 @@
 // [START serviceextensions_plugin_hmac_authcookie]
 #include <openssl/hmac.h>
 
-#include <iomanip>
-#include <sstream>
 #include <string>
 
+#include "absl/strings/escaping.h"
 #include "absl/strings/str_split.h"
+#include "absl/time/clock.h"
+#include "absl/time/time.h"
 #include "proxy_wasm_intrinsics.h"
+#include "re2/re2.h"
 
 // Replace with your desired secret key.
 const std::string kSecretKey = "your_secret_key";
 
+class MyRootContext : public RootContext {
+ public:
+  explicit MyRootContext(uint32_t id, std::string_view root_id)
+      : RootContext(id, root_id) {}
+
+  bool onConfigure(size_t) override {
+    // Regex for matching IPs on format like 127.0.0.1.
+    ip_match.emplace("^(?:[0-9]{1,3}\\.){3}[0-9]{1,3}$");
+    return ip_match->ok();
+  }
+
+  std::optional<re2::RE2> ip_match;
+};
+
+// Validates the HMAC HTTP cookie performing the following steps:
+//
+// 1. Obtains the client IP address and rejects the request if it is not
+// present.
+// 2. Obtains the HTTP cookie and rejects the request if it is not present.
+// 3. Verifies that the HMAC hash of the cookie matches its payload, rejecting
+// the request if there is no match.
+// 4. Ensures that the client IP address matches the IP in the cookie payload,
+// and that the current time is earlier than the expiration time specified in
+// the cookie payload.
 class MyHttpContext : public Context {
  public:
-  explicit MyHttpContext(uint32_t id, RootContext* root) : Context(id, root) {}
+  explicit MyHttpContext(uint32_t id, RootContext* root)
+      : Context(id, root), root_(static_cast<MyRootContext*>(root)) {}
 
   FilterHeadersStatus onRequestHeaders(uint32_t headers,
                                        bool end_of_stream) override {
-    const auto token = getTokenFromCookie();
+    const std::optional<std::string> ip = getClientIp();
+    if (!ip.has_value()) {
+      LOG_INFO("Access forbidden - missing client IP.");
+      sendLocalResponse(403, "", "Access forbidden - missing client IP.\n", {});
+      return FilterHeadersStatus::ContinueAndEndStream;
+    }
+
+    const std::optional<std::string> token = getAuthorizationCookie();
     if (!token.has_value()) {
       LOG_INFO("Access forbidden - missing HMAC cookie.");
       sendLocalResponse(403, "", "Access forbidden - missing HMAC cookie.\n",
                         {});
       return FilterHeadersStatus::ContinueAndEndStream;
     }
 
-    const auto path = getRequestHeader(":path")->toString();
-    if (computeHmacSignature(path) != token.value()) {
+    const std::optional<std::pair<std::string, std::string>> payload_and_hash =
+        parseAuthorizationCookie(token.value());
```

Comment:
optional: can dereference using `*token` instead of `token.value()`, since `has_value()` has been checked earlier. `*token` is shorter and doesn't have behavior that depends on build mode (i.e. whether or not to throw an exception)